llvm/lib/CodeGen/ShrinkWrap.cpp

```diff
@@ -957,6 +957,15 @@ bool ShrinkWrap::runOnMachineFunction(MachineFunction &MF) {
   if (!ArePointsInteresting())
     return Changed;
 
+  const TargetLowering *TLI = MF.getSubtarget().getTargetLowering();
```

Comment:
Suggest placing this whole logic in `performShrinkWrapping` and/or `postShrinkWrapping`.
The test `TLI->isStackProbeInstrDefinedFlagRegLiveIn(Save)`  can quickly discard a block as a candidate for prologue/epilogue, so by placing it inside those functions we can avoid iterating over all instructions in a candidate block.

Reply:
In `performShrinkWrapping` we need to calculate `StackAddressUsed` for each block, regardless of whether the test `TLI->isStackProbeInstrDefinedFlagRegLiveIn()` returns true or false, so it seems that we have to iterate through instructions.

The current approach - in which the test is placed after all the analysis has been done - is relatively simple and safe - we don't have to worry about missing this test in some code paths in `performShrinkWrapping` or `postShrinkWrapping`.

Reply:
If my reading of `performShrinkWrapping` is correct, the `StackAddressUsed` is conservatively set for blocks that are not suitable for a prologue, not necessarily meaning they computed an SP-based address.
Thus we can set this variable for blocks with NZCV live-in.

Anyway, looking a bit more, I now think the proper place to fix this issue is `TargetFrameLowering::canUseAsPrologue`, so I hopefully fixed it there and I also found another bug:

https://github.com/llvm/llvm-project/pull/81878
https://github.com/llvm/llvm-project/pull/81879

Reply:
> Thus we can set this variable for blocks with NZCV live-in

A block with NZCV live-in wouldn't necessarily be selected for prologue insertion, so `StackAddressUsed` still need to be calculated by iterating through and examining each instruction.